// @skits/react-cursor/lib/cjs/Cursor.component.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cursor = void 0;
var react_1 = __importStar(require("react"));
var Cursor_1 = __importDefault(require("./Cursor"));
var Cursor = function (_a) {
    var _b = _a.size, size = _b === void 0 ? 40 : _b, _c = _a.opacity, opacity = _c === void 0 ? 1 : _c, _d = _a.color, color = _d === void 0 ? 'red' : _d, _e = _a.radius, radius = _e === void 0 ? '50%' : _e, _f = _a.zIndex, zIndex = _f === void 0 ? 999999 : _f, className = _a.className, onMouseDown = _a.onMouseDown, onMouseUp = _a.onMouseUp;
    var cursorElement = (0, react_1.useRef)(null);
    var cursorClass = (0, react_1.useRef)();
    (0, react_1.useEffect)(function () {
        if (cursorElement.current) {
            cursorClass.current = new Cursor_1.default(cursorElement.current, onMouseDown, onMouseUp);
        }
        return function () {
            if (cursorClass.current) {
                cursorClass.current.stop();
            }
        };
    }, [onMouseDown, onMouseUp]);
    var cursorStyles = (0, react_1.useMemo)(function () { return ({
        cursor: {
            position: 'fixed',
            pointerEvents: 'none',
            transformOrigin: 'center',
            transform: 'translate3d(-50%, -50%, 0)',
            willChange: 'transform',
            zIndex: zIndex,
        },
        shape: {
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            width: "".concat(size, "px"),
            height: "".concat(size, "px"),
            backgroundColor: color,
            borderRadius: radius,
            opacity: opacity,
            transition: 'all 0.4s cubic-bezier(0.19, 1, 0.22, 1)',
            backfaceVisibility: 'hidden',
            willChange: 'transform, width, height, opacity',
        },
    }); }, [color, opacity, radius, size, zIndex]);
    return (react_1.default.createElement("div", { ref: cursorElement, style: cursorStyles.cursor, className: "skits-cursor ".concat(className) },
        react_1.default.createElement("div", { style: cursorStyles.shape, className: "skits-cursor__shape" })));
};
exports.Cursor = Cursor;
