// @skits/react-cursor/lib/esm/Cursor.component.js
import React, { useEffect, useMemo, useRef } from 'react';
import CursorClass from './Cursor';
export var Cursor = function (_a) {
    var _b = _a.size, size = _b === void 0 ? 40 : _b, _c = _a.opacity, opacity = _c === void 0 ? 1 : _c, _d = _a.color, color = _d === void 0 ? 'red' : _d, _e = _a.radius, radius = _e === void 0 ? '50%' : _e, _f = _a.zIndex, zIndex = _f === void 0 ? 999999 : _f, className = _a.className, onMouseDown = _a.onMouseDown, onMouseUp = _a.onMouseUp;
    var cursorElement = useRef(null);
    var cursorClass = useRef();
    useEffect(function () {
        if (cursorElement.current) {
            cursorClass.current = new CursorClass(cursorElement.current, onMouseDown, onMouseUp);
        }
        return function () {
            if (cursorClass.current) {
                cursorClass.current.stop();
            }
        };
    }, [onMouseDown, onMouseUp]);
    var cursorStyles = useMemo(function () { return ({
        cursor: {
            position: 'fixed',
            pointerEvents: 'none',
            transformOrigin: 'center',
            transform: 'translate3d(-50%, -50%, 0)',
            willChange: 'transform',
            zIndex: zIndex,
        },
        shape: {
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            width: "".concat(size, "px"),
            height: "".concat(size, "px"),
            backgroundColor: color,
            borderRadius: radius,
            opacity: opacity,
            transition: 'all 0.4s cubic-bezier(0.19, 1, 0.22, 1)',
            backfaceVisibility: 'hidden',
            willChange: 'transform, width, height, opacity',
        },
    }); }, [color, opacity, radius, size, zIndex]);
    return (React.createElement("div", { ref: cursorElement, style: cursorStyles.cursor, className: "skits-cursor ".concat(className) },
        React.createElement("div", { style: cursorStyles.shape, className: "skits-cursor__shape" })));
};
